### PR TITLE
Add index_copy() operator

### PR DESCRIPTION
**Pull request by aksnzhy**
## Description ##
This is a PR to add a index_copy() tensor operator, which will be used in the DGL project.

`output = mx.nd.contrib.index_copy(old_tensor, index, new_tensor)`

This op copies the elements of a `new_tensor` into the `old_tensor` by 
selecting the indices in the order given in `index`. The output will be a new tensor 
contains the rest elements of old tensor and the copied elements of new tensor. 
For example, if `index[i] == j`, then the `i`th row of `new_tensor` is copied to the 
`j`th row of output.

The `index` must be a vector and it must have the same size with the `0`th dimimention of 
`new_tensor`. Also, the `0`th dimimention of old_tensor must `>=` the `0`th dimimention of 
`new_tensor`, or an error will be raised.

Examples:

```
x = mx.nd.zeros((5,3))
t = mx.nd.array([[1,2,3],[4,5,6],[7,8,9]])
index = mx.nd.array([0,4,2])

mx.nd.contrib.index_copy(x, index, t)

[[1. 2. 3.]
 [0. 0. 0.]
 [7. 8. 9.]
 [0. 0. 0.]
 [4. 5. 6.]]
<NDArray 5x3 @cpu(0)>
```

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [x] Add index_copy-inl.h, index_copy.cc, and index_copy.cu for index_copy operator.
- [x] Add test_index_copy() in Python unitest.

### CODE REVIEW

**Comment by sandeep-krishnamurthy** [comment]
@aksnzhy - Thanks for your contributions. 
Few initial queries
1. Should we also take `axis` or `dim` for doing the index copy?
2. This does not have GPU implementation?

Also, we follow a practice of creating a JIRA issue for anything significant contributions. Please create JIRA for this task here - https://issues.apache.org/jira/projects/MXNET/issues/

@samskalicky - Can you please help take a look at this new operator?

**Comment by zheng-da** [comment]
There is a GPU implementation.
https://github.com/apache/incubator-mxnet/pull/12810/files#diff-c82104024d2308a03ff08648250105f1R30

File: src/operator/contrib/index_copy-inl.h [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+namespace op {
+
+struct copy_all {
+  template<typename DType>
+  MSHADOW_XINLINE static void Map(int i,
+                                  DType* old_tensor,
+                                  DType* out_tensor) {
+    *(out_tensor + i) = *(old_tensor + i);
+  }
+};
```
Comment (zheng-da) [inline src/operator/contrib/index_copy-inl.h]: We should have a copy_all implementation. It's called identity? @eric-haibin-lin

**Comment by zheng-da** [comment]
Should we put it in contrib? this operator should be pretty mature.

**Comment by zheng-da** [comment]
@aksnzhy could you please implement backward computation?

**Comment by aksnzhy** [comment]
> @aksnzhy - Thanks for your contributions.
> Few initial queries
> 
> 1. Should we also take `axis` or `dim` for doing the index copy?
> 2. This does not have GPU implementation?
> 
> Also, we follow a practice of creating a JIRA issue for anything significant contributions. Please create JIRA for this task here - https://issues.apache.org/jira/projects/MXNET/issues/
> 
> @samskalicky - Can you please help take a look at this new operator?

1. In this op, we assume that the `dim=0` for `t`

**Comment by aksnzhy** [comment]
> @aksnzhy - Thanks for your contributions.
> Few initial queries
> 
> 1. Should we also take `axis` or `dim` for doing the index copy?
> 2. This does not have GPU implementation?
> 
> Also, we follow a practice of creating a JIRA issue for anything significant contributions. Please create JIRA for this task here - https://issues.apache.org/jira/projects/MXNET/issues/
> 
> @samskalicky - Can you please help take a look at this new operator?

1. This is op, we don't need to take `axis` or `dim`, and the operator will check the shape matching automatically.
2. We have GPU implementation in this PR.

Thanks!

File: src/operator/contrib/index_copy-inl.h (hunk 1 above)
Comment (eric-haibin-lin) [reply src/operator/contrib/index_copy-inl.h]: There is `mxnet_op::copy(ctx.get_stream<xpu>(), outputs[0], inputs[0]);`. See `static void IdentityCompute` in tensor/elemwise_unary_op.h

File: src/operator/contrib/index_copy.cc [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+ [7. 8. 9.]
+ [0. 0. 0.]
+ [4. 5. 6.]]
+<NDArray 5x3 @cpu(0)>
+
+)code" ADD_FILELINE)
+.set_num_inputs(3)
+.set_num_outputs(1)
+.set_attr<nnvm::FInferShape>("FInferShape", IndexCopyShape)
+.set_attr<nnvm::FInferType>("FInferType", IndexCopyType)
```
[the remark this answers is not in the document]
Comment (aksnzhy) [reply src/operator/contrib/index_copy.cc]: Change it to ElemwiseType<3, 1>

**Pull request merged**